python/pyarrow/_dataset.pyx

```diff
@@ -270,19 +425,20 @@ cdef class FileSystemDataset(Dataset):
         CFileSystemDataset* filesystem_dataset
 
     def __init__(self, paths_or_selector, schema=None, format=None,
-                 filesystem=None, partitions=None, root_partition=None):
+                 filesystem=None, partitions=None, root_partition=_true):
```

Comment:
But an expression is not mutable? (from the cython/python side) In which case that shouldn't be a problem like lists are.

But, I agree that having `None` as default here is idiomatic and would be fine to keep.

Reply:
Correct, expressions are not mutable, until they are because of a refactor. The point is, this is an anti-pattern in python due to odd side-effect of default argument behavior, on top of being non-idiomatic.

Reply:
Okay, I'll coerce `None` to `_true` in the function body. If you'd prefer that we don't coerce `None` expressions to `_true` in python at all, we'll need to expose and use C++ overloads which don't accept an expression (so the coercion can happen in C++), but that seems less preferred to me.

Reply:
It's fine to coerce in python.